### ISSUE TITLE
Unit tests fail locally after pulling latest changes

### ISSUE THREAD

**Issue by WeilerP** [theislab/cellrank#656]
## Description

After pulling the latest changes, the tests did not pass for me anymore when running

```bash
tox -e py38-linux
```

`tox` was using incompatible versions of `umap-learn` and `numba` (the same issue as in theislab/scanpy#1756 or theislab/scvelo#387) and a wrong version for `wot`. The issue with `wot` was

```tb
FAILED tests/test_external.py::TestWOTKernel::test_cost_matrices[good_shape] - TypeError: compute_transport_map() got an unexpected keyword argument 'cost_matrix'
FAILED tests/test_external.py::TestWOTKernel::test_last_time_point[connectivities] - TypeError: compute_transport_map() got an unexpected keyword argument 'cost_matrix'
FAILED tests/test_external.py::TestWOTKernel::test_no_connectivities - TypeError: compute_transport_map() got an unexpected keyword argument 'cost_matrix'
FAILED tests/test_external.py::TestWOTKernel::test_growth_rates[5] - TypeError: compute_transport_map() got an unexpected keyword argument 'cost_matrix'
FAILED tests/test_external.py::TestWOTKernel::test_normal_run - TypeError: compute_transport_map() got an unexpected keyword argument 'cost_matrix'
FAILED tests/test_external.py::TestWOTKernel::test_cost_matrices[X_pca] - TypeError: compute_transport_map() got an unexpected keyword argument 'cost_matrix'
FAILED tests/test_external.py::TestWOTKernel::test_last_time_point[diagonal] - TypeError: compute_transport_map() got an unexpected keyword argument 'cost_matrix'
FAILED tests/test_external.py::TestWOTKernel::test_growth_rates[3] - TypeError: compute_transport_map() got an unexpected keyword argument 'cost_matrix'
FAILED tests/test_external.py::TestWOTKernel::test_copy - TypeError: compute_transport_map() got an unexpected keyword argument 'cost_matrix'
FAILED tests/test_external.py::TestWOTKernel::test_cost_matrices[Ms] - TypeError: compute_transport_map() got an unexpected keyword argument 'cost_matrix'
FAILED tests/test_external.py::TestWOTKernel::test_last_time_point[uniform] - TypeError: compute_transport_map() got an unexpected keyword argument 'cost_matrix'
FAILED tests/test_external.py::TestWOTKernel::test_cost_matrices[None] - TypeError: compute_transport_map() got an unexpected keyword argument 'cost_matrix'
```

The keyword argument `cost_matrix` is not yet included in any tagged version of `wot`. I had to manually run 

```
pip install -e '.[dev]'
pip install git+https://github.com/broadinstitute/wot
```

to make the tests pass again. On the [latest CI run](https://github.com/theislab/cellrank/runs/2985828903?check_suite_focus=true) this morning, `wot==1.0.8.post2` was installed and the above mentioned tests skipped. The CI run triggered by #653 installed the correct, i.e. latest, version of `wot`.

#### Versions:

> cellrank==1.4.0+g1a71f63 scanpy==1.8.0 anndata==0.7.6 numpy==1.21.0 numba==0.52.0 scipy==1.7.0 pandas==1.2.5 pygpcca==1.0.2 scikit-learn==0.24.2 statsmodels==0.12.2 python-igraph==0.9.6 scvelo==0.2.3 pygam==0.8.0 matplotlib==3.4.2 seaborn==0.11.1

**Comment by michalk8** [comment]
> tox was using incompatible versions of umap-learn and numba (the same issue as in theislab/scanpy#1756 or theislab/scvelo#387) and a wrong version for wot. The issue with wot was

Strange, `umap-learn` should be only problem for `3.9` and the correct version of `wot` is on the `dev` branch (though not on master: https://github.com/theislab/cellrank/blob/dev/tox.ini#L100), will take a look.

**Comment by WeilerP** [comment]
Does `tox -e py38-linux` recheck the installed packages and down-/upgrade them if needed?

**Comment by michalk8** [comment]
> Does tox -e py38-linux recheck the installed packages and down-/upgrade them if needed?

Afaik not yet (see [here](https://github.com/tox-dev/tox/issues/149), it's supposedly on the `rewrite` branch), you'd have to run `tox -e <env> --recreate`.

**Comment by WeilerP** [comment]
> Strange, `umap-learn` should be only problem for `3.9` and the correct version of `wot` is on the `dev` branch (though not on master: https://github.com/theislab/cellrank/blob/dev/tox.ini#L100), will take a look. On `master`, `wot` is still being skipped.

Re the `wot` part: Doesn't this mean that after running `pip install 'cellrank[external]'` the `WOTKernel` cannot be used since `wot<=1.0.8.post2` is installed ATM?

**Comment by michalk8** [comment]
> Re the wot part: Doesn't this mean that after running pip install 'cellrank[external]' the WOTKernel cannot be used since wot<=1.0.8.post2 is installed ATM?

Yes, that's why in [docs](https://cellrank.readthedocs.io/en/stable/api/cellrank.external.kernels.WOTKernel.html#cellrank.external.kernels.WOTKernel) we have the infor that it should be installed from `git`. In general, you can't specify `git` requirements in `setup.py`.

**Comment by WeilerP** [comment]
> Yes, that's why in [docs](https://cellrank.readthedocs.io/en/stable/api/cellrank.external.kernels.WOTKernel.html#cellrank.external.kernels.WOTKernel) we have the infor that it should be installed from `git`. In general, you can't specify `git` requirements in `setup.py`.

Yes, a workaround would be to specify the requirements in a separate file, e.g. `requirements_external.txt` via

```
statot>=0.0.14
POT
wot @ git+https://github.com/broadinstitute/wot
```

and then parse this file in `setup.py` similar as it is done for `"docs"` ATM. Might be possible to define a single `requirements.yml` file and read the different packages from there.

**Comment by michalk8** [comment]
closed via #655 
let's continue the dev problems discussion in #628

**Comment by zhenzuo2** [comment]
When running 
```
wk.compute_transition_matrix(
    growth_iters=3, growth_rate_key="growth_rate_init", last_time_point="connectivities")
```
I got the following error message:

Computing transition matrix using Waddington optimal transport
WARNING: Unable to find HVGs in `adata.var['highly_variable']`. Using all genes
Using default cost matrices
Computing transport maps for `11` time pairs
0%
0/11 [00:00<?, ?time pair/s]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_63937/3737559325.py in <module>
----> 1 wk.compute_transition_matrix(
      2     growth_iters=3, growth_rate_key="growth_rate_init", last_time_point="connectivities")

/storage/singlecell/zz4/miniconda3/envs/mypython/lib/python3.8/site-packages/cellrank/external/kernels/_wot_kernel.py in compute_transition_matrix(self, cost_matrices, lambda1, lambda2, epsilon, growth_iters, solver, growth_rate_key, use_highly_variable, last_time_point, threshold, conn_kwargs, **kwargs)
    372             return self
    373 
--> 374         tmap = self._compute_pairwise_tmaps(
    375             adata,
    376             cost_matrices=cost_matrices,

/storage/singlecell/zz4/miniconda3/envs/mypython/lib/python3.8/site-packages/cellrank/external/kernels/_wot_kernel.py in _compute_pairwise_tmaps(self, adata, cost_matrices, solver, growth_rate_field, **kwargs)
    429         )
    430         for tpair, cost_matrix in tqdm(cost_matrices.items(), unit="time pair"):
--> 431             tmap: Optional[AnnData] = self._ot_model.compute_transport_map(
    432                 *tpair, cost_matrix=cost_matrix
    433             )

TypeError: compute_transport_map() got an unexpected keyword argument 'cost_matrix'

Is this related to this problem?

_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       1_gnu    conda-forge
alabaster                 0.7.12                     py_0    conda-forge
anndata                   0.7.6            py38h578d9bd_0    conda-forge
anyio                     3.3.0            py38h578d9bd_0    conda-forge
argon2-cffi               20.1.0           py38h497a2fe_2    conda-forge
async_generator           1.10                       py_0    conda-forge
attrs                     21.2.0             pyhd8ed1ab_0    conda-forge
babel                     2.9.1              pyh44b312d_0    conda-forge
backcall                  0.2.0              pyh9f0ad1d_0    conda-forge
backports                 1.0                        py_2    conda-forge
backports.functools_lru_cache 1.6.4              pyhd8ed1ab_0    conda-forge
bleach                    4.0.0              pyhd8ed1ab_0    conda-forge
blosc                     1.21.0               h9c3ff4c_0    conda-forge
brotlipy                  0.7.0           py38h497a2fe_1001    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
c-ares                    1.17.2               h7f98852_0    conda-forge
ca-certificates           2021.5.30            ha878542_0    conda-forge
cached-property           1.5.2                hd8ed1ab_1    conda-forge
cached_property           1.5.2              pyha770c72_1    conda-forge
cellrank                  1.5.0              pyhdfd78af_0    bioconda
cellrank-krylov           1.5.0              pyh1e54042_0    bioconda
certifi                   2021.5.30        py38h578d9bd_0    conda-forge
cffi                      1.14.6           py38ha65f79e_0    conda-forge
chardet                   4.0.0            py38h578d9bd_1    conda-forge
charset-normalizer        2.0.0              pyhd8ed1ab_0    conda-forge
click                     8.0.1            py38h578d9bd_0    conda-forge
colorama                  0.4.4              pyh9f0ad1d_0    conda-forge
cryptography              3.4.7            py38ha5dfef3_0    conda-forge
cycler                    0.10.0                     py_2    conda-forge
cython                    0.29.24                  pypi_0    pypi
debugpy                   1.4.1            py38h709712a_0    conda-forge
decorator                 5.0.9              pyhd8ed1ab_0    conda-forge
defusedxml                0.7.1              pyhd8ed1ab_0    conda-forge
docrep                    0.3.2              pyh44b312d_0    conda-forge
docutils                  0.17.1           py38h578d9bd_0    conda-forge
dunamai                   1.6.0              pyhd8ed1ab_0    conda-forge
entrypoints               0.3             pyhd8ed1ab_1003    conda-forge
freetype                  2.10.4               h0708190_1    conda-forge
future                    0.18.2           py38h578d9bd_3    conda-forge
get_version               3.5                pyhd8ed1ab_0    conda-forge
h5py                      3.3.0           nompi_py38h9915d05_100    conda-forge
hdf5                      1.10.6          nompi_h6a2412b_1114    conda-forge
hypre                     2.21.0          mpi_openmpi_h3e9f1a4_2    conda-forge
idna                      3.1                pyhd3deb0d_0    conda-forge
imagesize                 1.2.0                      py_0    conda-forge
importlib-metadata        4.6.4            py38h578d9bd_0    conda-forge
importlib_metadata        4.6.4                hd8ed1ab_0    conda-forge
iprogress                 0.4                        py_0    conda-forge
ipykernel                 6.2.0            py38he5a9106_0    conda-forge
ipython                   7.26.0           py38he5a9106_0    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
ipywidgets                7.6.3              pyhd3deb0d_0    conda-forge
jbig                      2.1               h7f98852_2003    conda-forge
jedi                      0.18.0           py38h578d9bd_2    conda-forge
jinja2                    3.0.1              pyhd8ed1ab_0    conda-forge
joblib                    1.0.1              pyhd8ed1ab_0    conda-forge
jpeg                      9d                   h36c2ea0_0    conda-forge
json5                     0.9.5              pyh9f0ad1d_0    conda-forge
jsonschema                3.2.0              pyhd8ed1ab_3    conda-forge
jupyter_client            7.0.1              pyhd8ed1ab_0    conda-forge
jupyter_core              4.7.1            py38h578d9bd_0    conda-forge
jupyter_server            1.10.2             pyhd8ed1ab_0    conda-forge
jupyterlab                3.1.8              pyhd8ed1ab_0    conda-forge
jupyterlab_pygments       0.1.2              pyh9f0ad1d_0    conda-forge
jupyterlab_server         2.7.2              pyhd8ed1ab_0    conda-forge
jupyterlab_widgets        1.0.0              pyhd8ed1ab_1    conda-forge
kiwisolver                1.3.1            py38h1fd1430_1    conda-forge
krb5                      1.19.2               hcc1bbae_0    conda-forge
lcms2                     2.12                 hddcbb42_0    conda-forge
ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge
legacy-api-wrap           1.2                        py_0    conda-forge
lerc                      2.2.1                h9c3ff4c_0    conda-forge
libblas                   3.9.0           11_linux64_openblas    conda-forge
libcblas                  3.9.0           11_linux64_openblas    conda-forge
libcurl                   7.78.0               h2574ce0_0    conda-forge
libdeflate                1.7                  h7f98852_5    conda-forge
libedit                   3.1.20191231         he28a2e2_2    conda-forge
libev                     4.33                 h516909a_1    conda-forge
libffi                    3.3                  h58526e2_2    conda-forge
libgcc-ng                 11.1.0               hc902ee8_8    conda-forge
libgfortran-ng            11.1.0               h69a702a_8    conda-forge
libgfortran5              11.1.0               h6c583b3_8    conda-forge
libgomp                   11.1.0               hc902ee8_8    conda-forge
liblapack                 3.9.0           11_linux64_openblas    conda-forge
libllvm10                 10.0.1               he513fc3_3    conda-forge
libnghttp2                1.43.0               h812cca2_0    conda-forge
libopenblas               0.3.17          pthreads_h8fe5266_1    conda-forge
libpng                    1.6.37               h21135ba_2    conda-forge
libsodium                 1.0.18               h36c2ea0_1    conda-forge
libssh2                   1.9.0                ha56f1ee_6    conda-forge
libstdcxx-ng              11.1.0               h56837e0_8    conda-forge
libtiff                   4.3.0                hf544144_1    conda-forge
libwebp-base              1.2.1                h7f98852_0    conda-forge
llvmlite                  0.35.0           py38h4630a5e_1    conda-forge
loompy                    3.0.6                      py_0    conda-forge
lz4-c                     1.9.3                h9c3ff4c_1    conda-forge
lzo                       2.10              h516909a_1000    conda-forge
markupsafe                2.0.1            py38h497a2fe_0    conda-forge
matplotlib-base           3.4.3            py38hf4fb855_0    conda-forge
matplotlib-inline         0.1.2              pyhd8ed1ab_2    conda-forge
metis                     5.1.0             h58526e2_1006    conda-forge
mistune                   0.8.4           py38h497a2fe_1004    conda-forge
mock                      4.0.3            py38h578d9bd_1    conda-forge
mpi                       1.0                     openmpi    conda-forge
mpi4py                    3.1.1            py38h3e8e7aa_0    conda-forge
mumps-include             5.2.1               ha770c72_10    conda-forge
mumps-mpi                 5.2.1               h4f59553_10    conda-forge
natsort                   7.1.1              pyhd8ed1ab_0    conda-forge
nbclassic                 0.3.1              pyhd8ed1ab_1    conda-forge
nbclient                  0.5.4              pyhd8ed1ab_0    conda-forge
nbconvert                 6.1.0            py38h578d9bd_0    conda-forge
nbformat                  5.1.3              pyhd8ed1ab_0    conda-forge
ncurses                   6.2                  h58526e2_4    conda-forge
nest-asyncio              1.5.1              pyhd8ed1ab_0    conda-forge
networkx                  2.5                        py_0    conda-forge
notebook                  6.4.3              pyha770c72_0    conda-forge
numba                     0.52.0           py38h51da96c_0    conda-forge
numexpr                   2.7.3            py38h51da96c_0    conda-forge
numpy                     1.21.2           py38he2449b9_0    conda-forge
numpy_groupies            0.9.13             pyh9f0ad1d_1    conda-forge
olefile                   0.46               pyh9f0ad1d_1    conda-forge
openjpeg                  2.4.0                hb52868f_1    conda-forge
openmpi                   4.1.1                hbfc84c5_0    conda-forge
openssl                   1.1.1l               h7f98852_0    conda-forge
packaging                 21.0               pyhd8ed1ab_0    conda-forge
pandas                    1.3.2            py38h43a58ef_0    conda-forge
pandoc                    2.14.2               h7f98852_0    conda-forge
pandocfilters             1.4.2                      py_1    conda-forge
parmetis                  4.0.3             he9a3056_1005    conda-forge
parso                     0.8.2              pyhd8ed1ab_0    conda-forge
patsy                     0.5.1                      py_0    conda-forge
petsc                     3.15.1          real_h568e158_100    conda-forge
petsc4py                  3.15.0          real_h68baad8_101    conda-forge
pexpect                   4.8.0              pyh9f0ad1d_2    conda-forge
pickleshare               0.7.5                   py_1003    conda-forge
pillow                    8.3.1            py38h8e6f84c_0    conda-forge
pip                       21.2.4             pyhd8ed1ab_0    conda-forge
pot                       0.7.0                    pypi_0    pypi
progressbar2              3.53.1             pyh9f0ad1d_0    conda-forge
prometheus_client         0.11.0             pyhd8ed1ab_0    conda-forge
prompt-toolkit            3.0.20             pyha770c72_0    conda-forge
ptscotch                  6.0.9                hedbc1fb_1    conda-forge
ptyprocess                0.7.0              pyhd3deb0d_0    conda-forge
pycparser                 2.20               pyh9f0ad1d_2    conda-forge
pygam                     0.8.0                      py_0    conda-forge
pygments                  2.10.0             pyhd8ed1ab_0    conda-forge
pygpcca                   1.0.2            py38h578d9bd_0    conda-forge
pynndescent               0.5.4              pyh6c4a22f_0    conda-forge
pyopenssl                 20.0.1             pyhd8ed1ab_0    conda-forge
pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge
pyrsistent                0.17.3           py38h497a2fe_2    conda-forge
pysocks                   1.7.1            py38h578d9bd_3    conda-forge
pytables                  3.6.1            py38hc386592_3    conda-forge
python                    3.8.10          h49503c6_1_cpython    conda-forge
python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge
python-utils              2.5.6              pyh44b312d_0    conda-forge
python_abi                3.8                      2_cp38    conda-forge
pytz                      2021.1             pyhd8ed1ab_0    conda-forge
pyzmq                     22.2.1           py38h2035c66_0    conda-forge
readline                  8.1                  h46c0cb4_0    conda-forge
requests                  2.26.0             pyhd8ed1ab_0    conda-forge
requests-unixsocket       0.2.0                      py_0    conda-forge
scalapack                 2.0.2             hc7fdd6d_1009    conda-forge
scanpy                    1.8.1              pyhd8ed1ab_0    conda-forge
scikit-learn              0.24.2           py38h1561384_1    conda-forge
scipy                     1.7.1            py38h56a6a73_0    conda-forge
scotch                    6.0.9                h3858553_1    conda-forge
scvelo                    0.2.4              pyhdfd78af_0    bioconda
seaborn                   0.11.2               hd8ed1ab_0    conda-forge
seaborn-base              0.11.2             pyhd8ed1ab_0    conda-forge
send2trash                1.8.0              pyhd8ed1ab_0    conda-forge
setuptools                57.4.0           py38h578d9bd_0    conda-forge
sinfo                     0.3.1                      py_0    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
slepc                     3.15.1          real_h6e7ff00_101    conda-forge
slepc4py                  3.15.1           py38h2df54b5_0    conda-forge
sniffio                   1.2.0            py38h578d9bd_1    conda-forge
snowballstemmer           2.1.0              pyhd8ed1ab_0    conda-forge
sphinx                    4.1.2              pyh6c4a22f_1    conda-forge
sphinxcontrib-applehelp   1.0.2                      py_0    conda-forge
sphinxcontrib-devhelp     1.0.2                      py_0    conda-forge
sphinxcontrib-htmlhelp    2.0.0              pyhd8ed1ab_0    conda-forge
sphinxcontrib-jsmath      1.0.1                      py_0    conda-forge
sphinxcontrib-qthelp      1.0.3                      py_0    conda-forge
sphinxcontrib-serializinghtml 1.1.5              pyhd8ed1ab_0    conda-forge
sqlite                    3.36.0               h9cd32fc_0    conda-forge
statot                    0.0.14                   pypi_0    pypi
statsmodels               0.12.2           py38h5c078b8_0    conda-forge
stdlib-list               0.7.0                      py_2    conda-forge
suitesparse               5.7.2                h7a0d4b7_0    conda-forge
superlu                   5.2.2                h16cfea0_0    conda-forge
tbb                       2020.2               h4bd325d_4    conda-forge
terminado                 0.11.1           py38h578d9bd_0    conda-forge
testpath                  0.5.0              pyhd8ed1ab_0    conda-forge
threadpoolctl             2.2.0              pyh8a188c0_0    conda-forge
tk                        8.6.11               h21135ba_0    conda-forge
tornado                   6.1              py38h497a2fe_1    conda-forge
tqdm                      4.62.2             pyhd8ed1ab_0    conda-forge
traitlets                 5.0.5                      py_0    conda-forge
typing_extensions         3.10.0.0           pyha770c72_0    conda-forge
umap-learn                0.4.6                    pypi_0    pypi
urllib3                   1.26.6             pyhd8ed1ab_0    conda-forge
wcwidth                   0.2.5              pyh9f0ad1d_2    conda-forge
webencodings              0.5.1                      py_1    conda-forge
websocket-client          0.57.0           py38h578d9bd_4    conda-forge
wheel                     0.37.0             pyhd8ed1ab_1    conda-forge
widgetsnbextension        3.5.1            py38h578d9bd_4    conda-forge
wot                       1.0.8.post3.dev8+gca5e94f          pypi_0    pypi
wrapt                     1.12.1           py38h497a2fe_3    conda-forge
xlrd                      1.2.0                    pypi_0    pypi
xz                        5.2.5                h516909a_1    conda-forge
yaml                      0.2.5                h516909a_0    conda-forge
zeromq                    4.3.4                h9c3ff4c_0    conda-forge
zipp                      3.5.0              pyhd8ed1ab_0    conda-forge
zlib                      1.2.11            h516909a_1010    conda-forge
zstd                      1.5.0                ha95c52a_0    conda-forge
